### PR TITLE
[NFC] Basics: add `SendableBox` and `TokenBucket` helpers

### PR DESCRIPTION
**Pull request by MaxDesiatov**
### Motivation:

When working with Swift Concurrency, Dispatch-based primitives such as locks and semaphores are not suitable. We should have `async`-friendly primitives that replace those.

### Modifications:

New `SendableBox` and `TokenBucket` actor types were added. I also took this as an opportunity to split `ConcurrencyHelpers.swift` into smaller files and move relevant files into the new `Basics/Concurrency` directory.

### Result:

`SendableBox` and `TokenBucket` are available for implementation and testing of features such as the new `async`-friendly `HTTPClient`, which is coming in a subsequent PR.

Depends on https://github.com/apple/swift-package-manager/pull/6058 and is a requirement of https://github.com/apple/swift-package-manager/pull/6022 for PR stacking purposes.

### CODE REVIEW

File: Sources/Basics/Concurrency/AsyncBox.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if swift(>=5.5.2)
+
+import struct Foundation.Date
+
+/// A `Sendable` storage that allows access from concurrently running tasks in an `async` closure.
+public actor AsyncBox<Value: Sendable> {
```
Comment (tomerd) [inline Sources/Basics/Concurrency/AsyncBox.swift:18]: curious, why is this even needed?

File: Sources/Basics/Concurrency/NSLock+Extensions.swift [hunk 2]
```diff
@@ -12,9 +12,9 @@
 
 import class Foundation.NSLock
 
-extension NSLock {
+public extension NSLock {
     /// Execute the given block while holding the lock.
-    public func withLock<T> (_ body: () throws -> T) rethrows -> T {
+    func withLock<T>(_ body: () throws -> T) rethrows -> T {
```
Comment (tomerd) [inline Sources/Basics/Concurrency/NSLock+Extensions.swift:17]: we prefer the visibility modifier on the functions not the extension

File: Sources/Basics/Concurrency/TokenBucket.swift [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#if swift(>=5.5.2)
+
+import DequeModule
+
+/// Type modeled after a "token bucket" pattern, which is similar to a semaphore, but is built with
+/// Swift Concurrency primitives.
+actor TokenBucket {
```
Comment (tomerd) [inline Sources/Basics/Concurrency/TokenBucket.swift:19]: does this need to public so it can be used by other modules?

**Review by tomerd — APPROVED** [review]

File: Sources/Basics/Concurrency/AsyncBox.swift (hunk 1 above)
Comment (MaxDesiatov) [reply Sources/Basics/Concurrency/AsyncBox.swift:18]: In `HTTPClient` tests we need a counter for checking how many times a closure was invoked and we need to assert on that counter after the closure has run. Previously this was done with a lock. Now with Swift Concurrency `async` closures can't capture an arbitrary variably, but they can capture references to actors.

File: Sources/Basics/Concurrency/TokenBucket.swift (hunk 3 above)
Comment (MaxDesiatov) [reply Sources/Basics/Concurrency/TokenBucket.swift]: Thanks, that's addressed now

**Review by neonichu — APPROVED** [review]

File: Sources/Basics/Concurrency/AsyncBox.swift (hunk 1 above)
Comment (MaxDesiatov) [reply Sources/Basics/Concurrency/AsyncBox.swift:18]: Ah, I remember the details now. It's a direct replacement for `ThreadSafeBox`.

Comment (tomerd) [reply Sources/Basics/Concurrency/AsyncBox.swift:18]: okay - maybe `SendableBox` then?

**Comment by MaxDesiatov** [comment]
For some reason GitHub automatically closed this PR when I merged the base branch, instead of retargeting it against `main` like it used to do. I also can't reopen it and I'm not sure why. I've re-created it as https://github.com/apple/swift-package-manager/pull/6085, but needs review approvals again unfortunately.

**Pull request closed without merging**